qiskit/algorithms/optimizers/steppable_optimizer.py

```diff
@@ -0,0 +1,304 @@
+
+
+@dataclass
+class TellData(ABC):
+    """Base class for argument type of :meth:`~.SteppableOptimizer.tell`.
+
+    Args:
+        eval_fun: Image of the function at :attr:`~.ask_data.x_fun`.
+        eval_jac: Image of the gradient-jacobian at :attr:`~.ask_data.x_jac`.
+
+    """
+
+    eval_fun: Union[float, List[float], None] = None
+    eval_jac: Union[POINT, List[POINT], None] = None
+
+
+@dataclass
+class OptimizerState:
+    """Base class representing the state of the optimizer.
+
+    Any variable that changes during the optimization should be stored in this dataclass.
+    """
+
+    x: POINT  # pylint: disable=invalid-name
+    fun: Optional[Callable[[POINT], float]]
+    jac: Optional[Callable[[POINT], POINT]]
+    nfev: Optional[int]
+    njev: Optional[int]
+    nit: Optional[int]
+
+
+class SteppableOptimizer(Optimizer):
+    """
+    Base class for a steppable optimizer.
+
+    This family of optimizers will be using the
+    `ask and tell interface
+    <https://optuna.readthedocs.io/en/stable/tutorial/20_recipes/009_ask_and_tell.html>`_.
+    When using this interface the user has to call :meth:`~.ask` in order to get information about
+    how to evaluate the fucntion (we are asking the optimizer about how to do the evaluation).
+    This information is typically the points at which we should evaluate the function next, but
+    depending on the optimizer it can also be about whether we should evaluate the function itself or
+    its gradient.
+    Once the function has been evaluated, the user calls the method :meth:`~..tell`
+    to tell the optimizer what the result of the function evaluation(s) is. The optimizer then
+    updates its state accordingly and the user can decide whether to stop the optimization process
+    or to repeat a step.
+
+    This interface is more customizable, and allows the user to have full control over the evaluation
+    of the function.
+
+    Examples:
+
+        An example where the evaluation of the function has a chance of failing. The user, with
+        specific knowledge about his function can catch this errors and handle them before passing
+        the result to the optimizer.
+
+        .. code-block:: python
+
+            import random
+            import numpy as np
+            from qiskit.algorithms.optimizers import GradientDescent
+
+            def objective(x):
+                if random.choice([True, False]):
+                    return None
+                else:
+                    return (np.linalg.norm(x) - 1) ** 2
+
+            def grad(x):
+                if random.choice([True, False]):
+                    return None
+                else:
+                    return 2 * (np.linalg.norm(x) - 1) * x / np.linalg.norm(x)
+
+
+            initial_point = np.random.normal(0, 1, size=(100,))
+
+            optimizer = GradientDescent(maxiter=20)
+            optimizer.start(x0=initial_point, fun=objective, jac=grad)
+
+            for _ in range(20):
```

Comment:
So if I understand things a call to `ask` followed by `tell` is defined as `step` or a single iteration. Now the loop here does 20 loops right - range(20). I guess I have to ask what maxiter does in relation to this and why they seem at first sight to be the same. If the loop was changed to while True would it end when maxiter was reached or is my loop determing that - or a combination?

Reply:
Yes, you are right that in this case it wouldn't make too much sense to write a loop like that.